src/api/habits/adapters/json.ts

```diff
@@ -9,24 +9,56 @@ const jsonHabitServiceFactory = (): HabitService => {
   const addHabit = (): Promise<string> =>
     Promise.resolve('adding habit in ze factorie');
 
-  const getHabits = (): Promise<GoalWithHabitHistory[]> =>
-    fetch('../data/habits.json')
-      .then((res) => res.json() as Promise<Response>)
-      .then((json) =>
-        json.map(({ habits, ...rest }) => ({
-          ...rest,
-          habits: habits.map(({ entries, frequency, ...restHabits }) => ({
-            ...restHabits,
-            frequency: frequency as Frequency,
-            entries: entries.map(({ completionDate, ...restEntries }) => ({
-              ...restEntries,
-              completionDate: new Date(completionDate),
-            })),
-          })),
-        }))
+  const parseJsonHabits = (json: string) => {
+    const jsonHabits = JSON.parse(json) as unknown as Response;
+
+    return jsonHabits.map(({ habits, ...rest }) => ({
+      ...rest,
+      habits: habits.map(({ entries, frequency, ...restHabits }) => ({
+        ...restHabits,
+        frequency: frequency as Frequency,
+        entries: entries.map(({ completionDate, ...restEntries }) => ({
+          ...restEntries,
+          completionDate: new Date(completionDate),
+        })),
+      })),
+    }));
+  };
+
+  const getHabits = (): Promise<GoalWithHabitHistory[]> => {
+    const localHabits = localStorage.getItem('habits');
+    console.log('localHabits', localHabits);
+    if (!localHabits) return Promise.resolve([]);
+
+    try {
+      return Promise.resolve(parseJsonHabits(localHabits));
+    } catch (error) {
+      console.error('Error fetching habits');
+      return Promise.reject(new Error('Failed to fetch habits'));
+    }
+  };
+
+  const saveHabits = (habits: GoalWithHabitHistory[]) => {
+    try {
+      localStorage.setItem('habits', JSON.stringify(habits));
+    } catch (error) {
+      return Promise.reject(
+        new Error('Failed to save habits to local storage.')
       );
+    }
+
+    return Promise.resolve('Saved habits successfully.');
+  };
+
+  // Save a default set of habits to local storage
+  const saveDefaultData = () => {
+    console.log('saving default data');
+    fetch('../data/habits.json')
```

Comment:
you may wish to return the promise from this, purely so that you have the option later on to await any of these actions to be completed

Reply:
I did consider that but wasn't sure if it's useful since this is not likely to be extended into further features and is probably going to be removed later once the app is launching.

But tbh for consistency and getting into good habits, I'll do it anyway :)